shared/pingpong/pingpong.go

```diff
@@ -1279,3 +1319,158 @@ func signTxn(signer *pingPongAccount, txn transactions.Transaction, cfg PpConfig
+		pps.latencyOuts = append(pps.latencyOuts, gzout)
+	} else {
+		bw := bufio.NewWriter(fout)
+		pps.latencyOuts = append(pps.latencyOuts, bw)
+	}
+	pps.sentTxid = make(chan txidSendTime, 20)
+	pps.latencyBlocks = make(chan bookkeeping.Block, 1)
+	go pps.txidLatency(ctx)
+	go pps.txidLatencyBlockWaiter(ctx, ac)
+}
+
+type txidSendTimeIndexed struct {
+	txidSendTime
+	index int
+}
+
+const txidLatencySampleSize = 10000
+
+// thread which handles measuring total send-to-commit latency
+func (pps *WorkerState) txidLatency(ctx context.Context) {
+	done := ctx.Done()
+	byTxid := make(map[string]txidSendTimeIndexed, txidLatencySampleSize)
+	txidList := make([]string, 0, txidLatencySampleSize)
+	out := pps.latencyOuts[len(pps.latencyOuts)-1]
+	for {
+		select {
+		case st := <-pps.sentTxid:
+			if len(txidList) < txidLatencySampleSize {
+				index := len(txidList)
+				txidList = append(txidList, st.txid)
+				byTxid[st.txid] = txidSendTimeIndexed{
+					st,
+					index,
+				}
+			} else {
+				// random replacement
+				evict := rand.Intn(len(txidList))
+				delete(byTxid, txidList[evict])
+				txidList[evict] = st.txid
+				byTxid[st.txid] = txidSendTimeIndexed{
+					st,
+					evict,
+				}
+			}
+		case bl := <-pps.latencyBlocks:
+			now := time.Now()
+			txns, err := bl.DecodePaysetFlat()
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "block[%d] payset err %v", bl.Round(), err)
+				return
+			}
+			for _, stxn := range txns {
+				txid := stxn.ID().String()
+				st, ok := byTxid[txid]
+				if ok {
+					dt := now.Sub(st.when)
+					fmt.Fprintf(out, "%d\n", dt.Nanoseconds())
+				}
+			}
+		case <-done:
+			return
+		}
+	}
+}
+
+type flusher interface {
+	Flush() error
+}
+
+func (pps *WorkerState) txidLatencyDone() {
+	for i := len(pps.latencyOuts); i >= 0; i-- {
+		xo := pps.latencyOuts[i]
+		if fl, ok := xo.(flusher); ok {
+			err := fl.Flush()
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "%s: v", pps.cfg.TotalLatencyOut, err)
+			}
+		}
+		if cl, ok := xo.(io.Closer); ok {
+			err := cl.Close()
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "%s: v", pps.cfg.TotalLatencyOut, err)
+			}
+		}
+	}
+}
+
+func (pps *WorkerState) txidLatencyBlockWaiter(ctx context.Context, ac *libgoal.Client) {
+	done := ctx.Done()
+restart:
+	// I wish Go had macros
+	select {
+	case <-done:
+		close(pps.latencyBlocks)
+		return
+	default:
+	}
+	st, err := ac.Status()
+	if err != nil {
+		select {
+		case <-done:
+			close(pps.latencyBlocks)
+			return
+		default:
+		}
+		fmt.Fprintf(os.Stderr, "block waiter st : %v", err)
+		time.Sleep(5 * time.Second)
+		goto restart
+	}
+	nextRound := st.LastRound + 1
+	for {
+		select {
+		case <-done:
+			close(pps.latencyBlocks)
+			return
+		default:
+		}
+		st, err = ac.WaitForRound(nextRound)
+		if err != nil {
+			select {
+			case <-done:
+				close(pps.latencyBlocks)
+				return
+			default:
+			}
+			fmt.Fprintf(os.Stderr, "block waiter w: %v", err)
+			time.Sleep(5 * time.Second)
+			goto restart
```

Comment:
Looks like this loop feeds blocks to the `latencyBlocks` handling, which in turn calls `time.Now` to figure out the latency from the recorded time to block time.

But, looking at this bit here, is it possible that this loop will be sleeping when goal publishes a new block for consumption? If that happens, the `time.Now` used for calculation would contain that delay, right?

Since this is error handling, I suspect that we don't really expect small temporary errors like that, but wanted to check anyhow.

Reply:
reduced the err restart time to 1 second; maybe the error condition will go away and we'll restart the API calls and not oversleep the round change.